zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -155,62 +157,103 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
             code_hash.expr(),
         );
 
-        // Setup first call's context.
-        for (field_tag, value) in [
-            (CallContextFieldTag::Depth, 1.expr()),
-            (CallContextFieldTag::CallerAddress, tx_caller_address.expr()),
-            (CallContextFieldTag::CalleeAddress, tx_callee_address.expr()),
-            (CallContextFieldTag::CallDataOffset, 0.expr()),
-            (
-                CallContextFieldTag::CallDataLength,
-                tx_call_data_length.expr(),
+        let is_empty_code_hash = IsEqualGadget::construct(
+            cb,
+            code_hash.expr(),
+            Word::random_linear_combine_expr(
+                (*EMPTY_HASH_LE).map(|byte| byte.expr()),
+                cb.power_of_randomness(),
             ),
-            (CallContextFieldTag::Value, tx_value.expr()),
-            (CallContextFieldTag::IsStatic, 0.expr()),
-            (CallContextFieldTag::LastCalleeId, 0.expr()),
-            (CallContextFieldTag::LastCalleeReturnDataOffset, 0.expr()),
-            (CallContextFieldTag::LastCalleeReturnDataLength, 0.expr()),
-            (CallContextFieldTag::IsRoot, 1.expr()),
-            (CallContextFieldTag::IsCreate, tx_is_create.expr()),
-            (CallContextFieldTag::CodeHash, code_hash.expr()),
-        ] {
-            cb.call_context_lookup(true.expr(), Some(call_id.expr()), field_tag, value);
-        }
+        );
+
+        cb.condition(is_empty_code_hash.expr(), |cb| {
+            cb.require_equal(
+                "Tx to account with empty code should be persistent",
+                reversion_info.is_persistent(),
+                1.expr(),
+            );
+            cb.require_equal(
+                "Go to EndTx when Tx to account with empty code",
```

Comment:
so this assumes tx to empty account must be in root call ? if there is internal call to empty account  will not go to `EndTx`

Reply:
No, call to account with empty code could happen in both case, but for internal call, it just need to restore the caller's context.

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/997293041e57bd727d38f24f7e976984b8ec4268/zkevm-circuits/src/evm_circuit/execution/callop.rs#L223-L245

Reply:
thanks, Callop handle it, another question :  Can we ensure Tx to account with empty code must be persistent ?  tx can failed due to many reason  i.e. Out of Gas, right ?
```
 cb.condition(is_empty_code_hash.expr(), |cb| {
            cb.require_equal(
                "Tx to account with empty code should be persistent",
                reversion_info.is_persistent(),
                1.expr(),
            );
```

Reply:
`BeginTx` handles the successful case, which doesn't consider invalid transaction. We need to handle  transactions with insufficient intrinsic gas or invalid signature, and treat it as no-op, but that not be our current focus.

Reply:
during I handle error case, I found even for step(opcode) failed,  the tx steps also consist of begin_tx . for example, begin_tx --> push1 --> push2(OOG occurs) --> end_tx ,  So I understand successful case is for final tx status , not the call scope status , right ?

Reply:
In this case, it's not an account with empty code, right?

Reply:
yes, I suspect in the case "n account with empty code", if error happens, the `is_persistent` is not true then . maybe better to an issue to warn us to check and add such test later.

Reply:
make sense, but I guess we could just left it to future implementer who implements invalid transaction. (invalid transaction I mean which never gets included in a L1 block, but we might need to support that in circuit in order to accept force-included transaction for censorship-resistance).

Reply:
sounds good!

Reply:
> make sense, but I guess we could just left it to future implementer who implements invalid transaction. (invalid transaction I mean which never gets included in a L1 block, but we might need to support that in circuit in order to accept force-included transaction for censorship-resistance).

Oh we are actually working on this, didn't think this would be generally useful but makes sense. We're only interested in being able to skip over transactions with invalid nonce/insufficient ETH (because these depend on the latest chain state), but we'll also do insufficient intrinsic gas/invalid signature if other people need it (though for forced transaction inclusion I think you could also just check this in the smart contract because you only need the tx data itself to verify these, but then you need extra solidity code as well to be able to do this so make sense to reuse circuits I guess).

Reply:
> Oh we are actually working on this, didn't think this would be generally useful but makes sense. We're only interested in being able to skip over transactions with invalid nonce/insufficient ETH

That's really nice! I think no matter what in the end we'd definitely like to move all transaction validation logic into circuit to reduce verifier cost (right now we could only validate all in smart contract).